### DISCUSSION TITLE
How to avoid race conditions in user requests that have mutations?

### DISCUSSION THREAD

**Discussion by hexwit** [graphql-java/graphql-java#2823] [Q&A]
Hi!

If multiple mutations executed against the same entity at the same time, but modifying different values - we can get into race conditions with unpredictable results. To avoid this, it is necessary to put mutations on the same entity (or request, or user) in a queue, and process them one by one. 
So the my question is how and where is better to do that? 
Is there any place where I can have access to request and mutations before they are going to be executed?

Or may be there exists another way how to handle race conditions?

**Comment by lqc** [comment]
Mutations in a single request are processed serially: https://spec.graphql.org/October2021/#sec-Mutation 

If you mean operations in different requests, then that depends on your backend/database layer and is not GraphQL specific.

**Reply by hexwit** [reply]
Within request - yes. I am saying about situation when user sends different requests at the same time. And there is no default mechanism to prevent races. 
TBH I didn't get message about GraphQL specific backend layer. My backend is spring + postgres. 

I am thinking that there should be some kind of orchestration, where it could be done. But not sure what exactly.
